{{cookiecutter.project_name}}/.github/workflows/release.yml

```diff
@@ -6,14 +6,16 @@
 
 name: Release
 on:
-  push:
-    branches: [main]
+  workflow_run:
+    workflows: ["tests"]
+    types:
+      - completed
 jobs:
   release:
     runs-on: ubuntu-latest
     concurrency: release
 
-    if: ${{ github.ref == 'refs/heads/main' }}
+    if: ${{ github.ref == 'refs/heads/main' && github.event.workflow_run.conclusion == 'success'}}
```

Comment:
So what will then happen in the workflow?

Before it did:
0) push to main
1) version bump 
2) commit (stops previous workflows -which leads to test "failing", which is slightly annoying)
3) run ci (tests + release)

I would ideally not want to double test, but also def. want the version bump to happen **always** regardless of tests.

So ideally I probably want:

0) push to main, no test are run except release 
1) version bump, but no release
2) push to main. tests will now be run
3) if tests pass (should probably be **test and documentation** if we have docs), create a release.

thoughts?

Reply:
> I would ideally not want to double test, but also def. want the version bump to happen always regardless of tests.

I'm not sure I understand that, can you clarify why? :-)

My thinking is that we use version control to be able to point to a given version.

Git: To point to versions during development.
Semantic versioning: To point to versions for users/during application. 

I would only want version bumps for stuff that we'd want users to install and use. If tests are not passing, I don't think that's the case.

Reply:
> I think that is completely unacceptable.

I think that's pretty strong wording for this ❤️ 

Yes, this situation could happen, but only if they're both installing from source. In that case, even if you're version bumping on every push to main, you could also have mismatched source code if the user edited the source code. 

For this reason, if the user is installing from source, I think the commit hash should be the source of versioning, not the version in the pyproject.toml.

That said, this is not something I want to argue about ❤️ 

I just know that the current situation, where we can release breaking software, is definitely not a good status quo.

Reply:
> Yes, this situation could happen, but only if they're both installing from source. 

You would only need one to do it.

> In that case, even if you're version bumping on every push to main, you could also have mismatched source code if the user edited the source code.

But you can't prevent that. I don't think it is comparable.

> I just know that the current situation, where we can release broken software, is definitely not a good status quo. I think this is a mergeable MVP, but feel free to modify it if version bumping, even with failing tests, is important to you 

but isn't that why we have tests and branch protection? I think the suggested solution is on par with the one we have and that this PR might suffer from having experiences a problem with it, while we might as well have experienced a problem the other way around. I think we can do better

So from the [docs](https://python-semantic-release.readthedocs.io/en/latest/automatic-releases/github-actions.html) it seems like we could use the [additional options](https://python-semantic-release.readthedocs.io/en/latest/index.html#cmd-common-options) using `--define` to overwrite the publish.

I think that would solve the problem without introducing a new one.

Reply:
Good points! I agree with most of what you said.

> But you can't prevent that. I don't think it is comparable.

Agreed, can't prevent it! 

But I think that, in the case where you're installing from source, you _shouldn't_ compare version numbers, you should compare git hashes. Comparing version numbers would be an antipattern, and bumping on every push to main would, in my eyes, reinforce that antipattern. 

> but isn't that why we have tests and branch protection? I think the suggested solution is on par with the one we have and that this PR might suffer from having experiences a problem with it, while we might as well have experienced a problem the other way around. I think we can do better

Currently, we can't prevent admins pushing to main because we have automatic version bumping, and that needs to push to main with an admin's PAT. This is the situation I would like to prevent. How we prevent that isn't too important to me :-)

Again, to my eyes, the current PR improves the current state, and from an "incremental improvement" standpoint, should be mergeable. That said, super open to a quick improvement.

> So from the [docs](https://python-semantic-release.readthedocs.io/en/latest/automatic-releases/github-actions.html) it seems like we could use the [additional options](https://python-semantic-release.readthedocs.io/en/latest/index.html#cmd-common-options) using --define to overwrite the publish.

Not sure I understand? Which configuration value could we `--define` to overwrite the publish? And how would that solve the problem?

Reply:
> But I think that, in the case where you're installing from source, you shouldn't compare version numbers, you should compare git hashes. Comparing version numbers would be an antipattern, and bumping on every push to main would, in my eyes, reinforce that antipattern.

That is a good point - assuming  `you should compare git hashes.` - which is not an argument that I am entirely convinced by - but I think that I agree that this PR is slightly better than the current state - lets merge it in

> Not sure I understand? Which configuration value could we --define to overwrite the publish? And how would that solve the problem?

release to pypi in the .toml?

Reply:
> release to pypi in the .toml?

I don't think I understand your suggestion here. In which situation would you override whether to release to pypi in the `.toml`, and how would that solve the problem? :-)